src/Extend/Floodgate.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Extend;
+
+use Flarum\Extension\Extension;
+use Illuminate\Contracts\Container\Container;
+
+class Floodgate implements ExtenderInterface
+{
+    private $setFloodgates = [];
+    private $removeFloodgates = [];
+
+    /**
+     * Register a flood gate checker.
+     *
+     * @param string $name: The name of the flood gate checker. These must be unique:
+     *                      if you use the name of an existing checker it will be overriden.
+     * @param string[] $paths: An array of paths (e.g. /api/discussions) on which to apply this checker.
+     *                         Wildcards are supported.
+     * @param string[] $methods: An array of methods on which to apply this checker.
+     *                           Methods should be uppercase, wildcards are NOT supported.
+     * @param string[] $callback: A closure or invokable class that contains the logic of the flood gate.
+     *                            It should return one of:
+     *                              - `false`: This marks the request as NOT flooding. It overrides all other outputs
+     *                              - `true`: This marks the request as flooding.
+     *                            All other outputs will be ignored.
+     * @return self
+     */
+    public function set(string $name, array $paths, array $methods, callable $callback)
```

Comment:
These arrays should allow for strings to be passed instead of one-element arrays (which would require no type hinting as PHP doesn't support string typings).

Or, continuing off my class idea, do something like

```php


(new Extend\Floodgate())
  ->add(
    Floodgate::build($name)
      ->routes($route1, $route2)
      ->methods($method1, $method2)
      ->handler($callback)
  )
```

Reply:
Mixed feelings on the API format. I like us having the freedom to assemble / store the actual flood gate objects as we see fit, that allows us to change things without breaking it later on.

Reply:
> allows us to change things without breaking it later on.

I feel like storing them as an array would make them break in the future more? If you want to change something, the array values would need to change. With this helper function, you can modify the internal code to adapt to the internally breaking changes ?

Reply:
We also don't have any other extenders that take an object. Although I do agree that at least internally, storing in an object makes more sense than storing in an array.

Reply:
But to follow up, the internal storage is just that, an implementation detail that we can tweak at any time since it's private API

Reply:
I think I agree with @datitisev ; question on my end is what the `Floodgate` is (not the extender I hope) and what `build` returns. Maybe an instance of `Throttle`? Do we have examples from eg Laravel on how they handle this?